### ISSUE TITLE
Requests being forgotten

### ISSUE THREAD

**Issue by busches** [mock-server/mockserver#1171]
**Describe the issue**
Similar to https://github.com/mock-server/mockserver/issues/794 when I upgrade from Mockserver 5.10.0 to 5.11.X, running `curl -X PUT "http://localhost:1080/mockserver/retrieve?type=REQUESTS" | grep "method" | wc -l` results in different numbers.

5.10.0 - 341
5.11.0 - 146
5.11.1 - 79
5.11.2 - 145
Latest `FROM mockserver/mockserver@sha256:29fdd72462a6b5c48de17c01d184b29327a1c96181de7bffac1defb5f7832986` - 103

From my application logs, I can see that the expectations/requests do exist, as my calls are succeeding, just the record of them is being wiped out.

**What you are trying to do**
Verify that calls were made to MockServer

**MockServer version**
See above.

**To Reproduce**

I'm running MockServer via Docker, loading my expectations via JSON files. These are my environment variables
```
      MOCKSERVER_MAX_LOG_ENTRIES: 20000 # Only added to try and help debug
      MOCKSERVER_INITIALIZATION_JSON_PATH: /config/initializerJson.json
```

**Expected behaviour**
The number of recorded expectations would match between 5.10.0 and higher

**MockServer Log**
```
2022-02-07 15:08:53 5.11.1 INFO loading JSON initialization file:
2022-02-07T15:08:54.011066200Z 
2022-02-07T15:08:54.011080700Z   /config/initializerJson.json
2022-02-07T15:08:54.011098300Z  
2022-02-07T15:08:54.134884200Z 2022-02-07 15:08:54 5.11.1 INFO processing JSON expectation 1 of 222
```
And it continues and creates them all.

**Comment by jamesdbloom** [comment]
The number of request being recorded has to be limited to avoid memory overload.  As this is held as part of a log of all events the number is not expected to be fixed.  The log size is fixed so depending of the type of interactions with MockServer the percentage of log events used for recording received request may change.  This design is important because it allows control over memory usage and allows the handling of a large number of incoming requests.

I will however investigate to see if there is something which could be done to tune the way this works.

It would be helpful it could provide a working code example showing how you produced those numbers then I would be able to invest and likely improve this prior to the next release.

**Comment by jamesdbloom** [comment]
I have tried to reproduce the low number of requests being recorded in your example using the following script:

```bash
#!/usr/bin/env bash

curl -v -s -X PUT "http://localhost:1080/mockserver/reset"

for counter in $(seq 1 1 5000); do
    # shellcheck disable=SC2059
    printf "\nnumber of requests sent: $counter"

    curl -s -X PUT 'http://localhost:1080/mockserver/expectation' -d "{
                        'httpRequest' : {
                        },
                        'httpOverrideForwardedRequest' : {
                          'responseModifier': {
                            'headers': {
                              'add': {
                                'x-proxy': ['mock-server']
                              }
                            }
                          },
                          'requestOverride': {
                            'socketAddress': {
                              'host' : 'mockserver_target',
                              'port' : 4567,
                              'scheme' : 'HTTPS'
                            }
                          },
                          'responseModifier': {
                            'headers': {
                              'add': {
                                'x-proxy': ['mock-server']
                              }
                            }
                          }
                        }
                      }" > /dev/null
    curl -s -X PUT 'http://localhost:1080/some/path' > /dev/null

    COUNT=$(curl -s -X PUT "http://localhost:1080/mockserver/retrieve?type=REQUESTS" | grep "method" | wc -l)
    printf " number of requests in log: $COUNT"
done
```

With this I find the number of recorded requests tops out at 2437 as the following output shows:

```bash
number of requests sent: 3097 number of requests in log:     2464
number of requests sent: 3098 number of requests in log:     2465
number of requests sent: 3099 number of requests in log:     2466
number of requests sent: 3100 number of requests in log:     2467
number of requests sent: 3101 number of requests in log:     2468
number of requests sent: 3102 number of requests in log:     2469
number of requests sent: 3103 number of requests in log:     2470
number of requests sent: 3104 number of requests in log:     2471
number of requests sent: 3105 number of requests in log:     2437
number of requests sent: 3106 number of requests in log:     2437
number of requests sent: 3107 number of requests in log:     2437
number of requests sent: 3108 number of requests in log:     2437
number of requests sent: 3109 number of requests in log:     2437
number of requests sent: 3110 number of requests in log:     2437
number of requests sent: 3111 number of requests in log:     2437
```

@busches  the number of recorded request also depends on the amount of available memory for the JVM, MockServer attempts to use what is available so in there is less memory free then it lowers the size of the log, this should ensure that is never causes an out of memory exception.  However, if you running MockServer in a JVM with little spare memory because either other applications are heavily using the memory or because the JVM didn't have much memory to start with the number of log entries will be lower and the number of recorded requests will also be lower.  Without this type of logic MockServer could use more memory then available and crash the JVM with an out of memory exception.

**Comment by busches** [comment]
I'm using the same machine running the same inputs via docker, so unless the new version uses a lot more memory, I can't understand the differences. They are repeatable each time. I'll see if I can specifically allocate fixed memory to the container and if that resolves it. Otherwise I'll see if I can create a reproduction for you.

**Comment by jamesdbloom** [comment]
But if any changes are made to the log events anyway in MockServer the number of recorded request could change.  I don't understand why this should be an issue, can you please explain what problem this is causing?

**Comment by busches** [comment]
Our use case is a cron job runs, then we verify all behavior, so it not remembering the original calls is the issue, as mocksever has no recollection of that call.

This is what perplexes me so, unless mocksever increased memory usage of each log, I'd expect no changes to the number remembered.

**Comment by jamesdbloom** [comment]
As the code changes the types of log events recorded are changed and the log events count is set based on free memory so if the log event number or size changes this could both impact the number of recorded requests. Certainly with most releases it is pretty likely one or both of those would change as the code evolves and is extended or bugs are fixed.

Perhaps you could increase the memory size for your instance of MockSever because when I tested on my laptop and inside a docker container I got well over 2000 recorded requests for some different scenarios.  I'm slightly confused why your examples seem so low in number.  If you can send me some code to reduce your numbers I'm happy to investigate further.

**Comment by busches** [comment]
Is there any easy way to see the max log events it will record? I did try
that override, 20k, and I got less events.

I will work on a reproduction, but as always, there are competing
priorities. Thanks for your help so far.

**Comment by busches** [comment]
Update, I did run into this again on 5.10.0 as well, as I had left around the 500 example.json files and instead of getting all 341 requests, I was capped at 284. So the problem exists in 5.10.0, but gets worse with 5.11.0. Increasing Memory/CPU/etc. does not change the limit, so it's something internal to MockServer it appears.

**Comment by MosheElisha** [comment]
@busches I am running MockServer 5.12.0 in Kubernetes and was faced with same issue. 
Configuring  `JVM_OPTIONS` with `-Xmx3g` didn't help but configuring `MOCKSERVER_LOG_LEVEL` to `ERROR` did help.

Perhaps you can try that.

**Comment by busches** [comment]
@MosheElisha we use the logs all the time to debug why a request isn't working, that's not a feasible workaround.

**Comment by MosheElisha** [comment]
@busches Makes sense. For us, usually everything is working and when a "verify" fails, we get all the requests in the "verify" response and usually are able to understand what happened.

**Comment by busches** [comment]
@MosheElisha that's one approach, but it doesn't offer the insight on why a request didn't match an expectation that the logs offer.

**Comment by jamesdbloom** [comment]
Thanks I'll take a detailed look at this

**Comment by jamesdbloom** [comment]
Why do you say the environment variable don't work, which ones don't work?

I can't seem to reproduce your issue.  With the following test I have managed to create 2859 expectations and have 2859 request in the log.  I stopped the test there because there didn't seem much point continuing.

I used the following bash script:

```bash
#!/usr/bin/env bash

docker stop mockserver > /dev/null 2>&1 || true
docker rm mockserver > /dev/null 2>&1 || true
docker run --rm --name mockserver -p 1080:1080 -e MOCKSERVER_LOG_LEVEL=OFF mockserver/mockserver:snapshot &
sleep 3
curl -v -s -X PUT "http://localhost:1080/mockserver/reset"

STATS=$(docker stats mockserver --no-stream --format "\tcpu: {{.CPUPerc}}  mem: {{.MemUsage}}")
for counter in $(seq 1 1 5000); do
  if [[ "$counter" == *0 ]]; then
    STATS=$(docker stats mockserver --no-stream --format "\tcpu: {{.CPUPerc}}  mem: {{.MemUsage}}")
  fi
  # shellcheck disable=SC2059
  printf "\nnumber of requests sent: %03d" "$counter"

  curl -s -X PUT 'http://localhost:1080/mockserver/expectation' -d "{
                        'httpRequest' : {
                        },
                        'httpOverrideForwardedRequest' : {
                          'responseModifier': {
                            'headers': {
                              'add': {
                                'x-proxy': ['mock-server']
                              }
                            }
                          },
                          'requestOverride': {
                            'socketAddress': {
                              'host' : 'mockserver_target',
                              'port' : 4567,
                              'scheme' : 'HTTPS'
                            }
                          },
                          'responseModifier': {
                            'headers': {
                              'add': {
                                'x-proxy': ['mock-server']
                              }
                            }
                          }
                        }
                      }" >/dev/null
  curl -s -X PUT 'http://localhost:1080/some/path' >/dev/null

  COUNT=$(curl -s -X PUT "http://localhost:1080/mockserver/retrieve?type=REQUESTS" | grep "method" | wc -l)
  EXPECTATIONS=$(curl -s -X PUT "http://localhost:1080/mockserver/retrieve?type=active_expectations" | grep "times" | wc -l)
  printf " number of requests in log: %03d" "$COUNT"
  printf " number of expectations: %03d" "$EXPECTATIONS"
  echo -n "$STATS"
done

cleanup() {
  docker stop mockserver || true
  docker rm mockserver || true
}

trap cleanup EXIT
```

Which generate the following results (truncated to fit in github):

```bash
./scripts/log_event_size_test.sh
*   Trying ::1:1080...
* Connected to localhost (::1) port 1080 (#0)
> PUT /mockserver/reset HTTP/1.1
> Host: localhost:1080
> User-Agent: curl/7.77.0
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< version: 5.13.1-SNAPSHOT
< connection: keep-alive
< content-length: 0
< 
* Connection #0 to host localhost left intact

number of requests sent: 001Loading JavaScript to validate ECMA262 regular expression in JsonSchema because java.util.regex package in Java does not match ECMA262
 number of requests in log: 001 number of expectations: 001	cpu: 0.38%  mem: 74.47MiB / 5.805GiB
number of requests sent: 002 number of requests in log: 002 number of expectations: 002	cpu: 0.38%  mem: 74.47MiB / 5.805GiB
number of requests sent: 003 number of requests in log: 003 number of expectations: 003	cpu: 0.38%  mem: 74.47MiB / 5.805GiB
number of requests sent: 004 number of requests in log: 004 number of expectations: 004	cpu: 0.38%  mem: 74.47MiB / 5.805GiB
number of requests sent: 005 number of requests in log: 005 number of expectations: 005	cpu: 0.38%  mem: 74.47MiB / 5.805GiB
number of requests sent: 006 number of requests in log: 006 number of expectations: 006	cpu: 0.38%  mem: 74.47MiB / 5.805GiB
...
number of requests sent: 203 number of requests in log: 203 number of expectations: 203	cpu: 0.36%  mem: 193.3MiB / 5.805GiB
number of requests sent: 204 number of requests in log: 204 number of expectations: 204	cpu: 0.36%  mem: 193.3MiB / 5.805GiB
number of requests sent: 205 number of requests in log: 205 number of expectations: 205	cpu: 0.36%  mem: 193.3MiB / 5.805GiB
number of requests sent: 206 number of requests in log: 206 number of expectations: 206	cpu: 0.36%  mem: 193.3MiB / 5.805GiB
number of requests sent: 207 number of requests in log: 207 number of expectations: 207	cpu: 0.36%  mem: 193.3MiB / 5.805GiB
number of requests sent: 208 number of requests in log: 208 number of expectations: 208	cpu: 0.36%  mem: 193.3MiB / 5.805GiB
number of requests sent: 209 number of requests in log: 209 number of expectations: 209	cpu: 0.36%  mem: 193.3MiB / 5.805GiB
number of requests sent: 210 number of requests in log: 210 number of expectations: 210	cpu: 0.33%  mem: 196.1MiB / 5.805GiB
number of requests sent: 211 number of requests in log: 211 number of expectations: 211	cpu: 0.33%  mem: 196.1MiB / 5.805GiB
number of requests sent: 212 number of requests in log: 212 number of expectations: 212	cpu: 0.33%  mem: 196.1MiB / 5.805GiB
...
number of requests sent: 349 number of requests in log: 349 number of expectations: 349	cpu: 0.37%  mem: 214.5MiB / 5.805GiB
number of requests sent: 350 number of requests in log: 350 number of expectations: 350	cpu: 0.39%  mem: 214.6MiB / 5.805GiB
number of requests sent: 351 number of requests in log: 351 number of expectations: 351	cpu: 0.39%  mem: 214.6MiB / 5.805GiB
number of requests sent: 352 number of requests in log: 352 number of expectations: 352	cpu: 0.39%  mem: 214.6MiB / 5.805GiB
number of requests sent: 353 number of requests in log: 353 number of expectations: 353	cpu: 0.39%  mem: 214.6MiB / 5.805GiB
number of requests sent: 354 number of requests in log: 354 number of expectations: 354	cpu: 0.39%  mem: 214.6MiB / 5.805GiB
number of requests sent: 355 number of requests in log: 355 number of expectations: 355	cpu: 0.39%  mem: 214.6MiB / 5.805GiB
number of requests sent: 356 number of requests in log: 356 number of expectations: 356	cpu: 0.39%  mem: 214.6MiB / 5.805GiB
number of requests sent: 357 number of requests in log: 357 number of expectations: 357	cpu: 0.39%  mem: 214.6MiB / 5.805GiB
number of requests sent: 358 number of requests in log: 358 number of expectations: 358	cpu: 0.39%  mem: 214.6MiB / 5.805GiB
number of requests sent: 359 number of requests in log: 359 number of expectations: 359	cpu: 0.39%  mem: 214.6MiB / 5.805GiB
number of requests sent: 360 number of requests in log: 360 number of expectations: 360	cpu: 0.37%  mem: 214.9MiB / 5.805GiB
...
number of requests sent: 656 number of requests in log: 656 number of expectations: 656	cpu: 0.30%  mem: 293.5MiB / 5.805GiB
number of requests sent: 657 number of requests in log: 657 number of expectations: 657	cpu: 0.30%  mem: 293.5MiB / 5.805GiB
number of requests sent: 658 number of requests in log: 658 number of expectations: 658	cpu: 0.30%  mem: 293.5MiB / 5.805GiB
number of requests sent: 659 number of requests in log: 659 number of expectations: 659	cpu: 0.30%  mem: 293.5MiB / 5.805GiB
number of requests sent: 660 number of requests in log: 660 number of expectations: 660	cpu: 0.23%  mem: 293.5MiB / 5.805GiB
number of requests sent: 661 number of requests in log: 661 number of expectations: 661	cpu: 0.23%  mem: 293.5MiB / 5.805GiB
number of requests sent: 662 number of requests in log: 662 number of expectations: 662	cpu: 0.23%  mem: 293.5MiB / 5.805GiB
...
number of requests sent: 1011 number of requests in log: 1011 number of expectations: 1011  cpu: 0.38%  mem: 421.7MiB / 5.805GiB
number of requests sent: 1012 number of requests in log: 1012 number of expectations: 1012  cpu: 0.38%  mem: 421.7MiB / 5.805GiB
number of requests sent: 1013 number of requests in log: 1013 number of expectations: 1013  cpu: 0.38%  mem: 421.7MiB / 5.805GiB
number of requests sent: 1014 number of requests in log: 1014 number of expectations: 1014  cpu: 0.38%  mem: 421.7MiB / 5.805GiB
number of requests sent: 1015 number of requests in log: 1015 number of expectations: 1015  cpu: 0.38%  mem: 421.7MiB / 5.805GiB
number of requests sent: 1016 number of requests in log: 1016 number of expectations: 1016  cpu: 0.38%  mem: 421.7MiB / 5.805GiB
number of requests sent: 1017 number of requests in log: 1017 number of expectations: 1017  cpu: 0.38%  mem: 421.7MiB / 5.805GiB
number of requests sent: 1018 number of requests in log: 1018 number of expectations: 1018  cpu: 0.38%  mem: 421.7MiB / 5.805GiB
...
number of requests sent: 1873 number of requests in log: 1873 number of expectations: 1873  cpu: 0.44%  mem: 891.5MiB / 5.805GiB
number of requests sent: 1874 number of requests in log: 1874 number of expectations: 1874  cpu: 0.44%  mem: 891.5MiB / 5.805GiB
number of requests sent: 1875 number of requests in log: 1875 number of expectations: 1875  cpu: 0.44%  mem: 891.5MiB / 5.805GiB
number of requests sent: 1876 number of requests in log: 1876 number of expectations: 1876  cpu: 0.44%  mem: 891.5MiB / 5.805GiB
number of requests sent: 1877 number of requests in log: 1877 number of expectations: 1877  cpu: 0.44%  mem: 891.5MiB / 5.805GiB
number of requests sent: 1878 number of requests in log: 1878 number of expectations: 1878  cpu: 0.44%  mem: 891.5MiB / 5.805GiB
number of requests sent: 1879 number of requests in log: 1879 number of expectations: 1879  cpu: 0.44%  mem: 891.5MiB / 5.805GiB
number of requests sent: 1880 number of requests in log: 1880 number of expectations: 1880  cpu: 19.04%  mem: 896.6MiB / 5.805GiB
number of requests sent: 1881 number of requests in log: 1881 number of expectations: 1881  cpu: 19.04%  mem: 896.6MiB / 5.805GiB
...
number of requests sent: 2100 number of requests in log: 2100 number of expectations: 2100  cpu: 0.46%  mem: 1002MiB / 5.805GiB
number of requests sent: 2101 number of requests in log: 2101 number of expectations: 2101  cpu: 0.46%  mem: 1002MiB / 5.805GiB
number of requests sent: 2102 number of requests in log: 2102 number of expectations: 2102  cpu: 0.46%  mem: 1002MiB / 5.805GiB
number of requests sent: 2103 number of requests in log: 2103 number of expectations: 2103  cpu: 0.46%  mem: 1002MiB / 5.805GiB
number of requests sent: 2104 number of requests in log: 2104 number of expectations: 2104  cpu: 0.46%  mem: 1002MiB / 5.805GiB
number of requests sent: 2105 number of requests in log: 2105 number of expectations: 2105  cpu: 0.46%  mem: 1002MiB / 5.805GiB
number of requests sent: 2106 number of requests in log: 2106 number of expectations: 2106  cpu: 0.46%  mem: 1002MiB / 5.805GiB
...
number of requests sent: 2487 number of requests in log: 2487 number of expectations: 2487  cpu: 0.43%  mem: 1.328GiB / 5.805GiB
number of requests sent: 2488 number of requests in log: 2488 number of expectations: 2488  cpu: 0.43%  mem: 1.328GiB / 5.805GiB
number of requests sent: 2489 number of requests in log: 2489 number of expectations: 2489  cpu: 0.43%  mem: 1.328GiB / 5.805GiB
number of requests sent: 2490 number of requests in log: 2490 number of expectations: 2490  cpu: 0.42%  mem: 1.345GiB / 5.805GiB
number of requests sent: 2491 number of requests in log: 2491 number of expectations: 2491  cpu: 0.42%  mem: 1.345GiB / 5.805GiB
number of requests sent: 2492 number of requests in log: 2492 number of expectations: 2492  cpu: 0.42%  mem: 1.345GiB / 5.805GiB
number of requests sent: 2493 number of requests in log: 2493 number of expectations: 2493  cpu: 0.42%  mem: 1.345GiB / 5.805GiB
number of requests sent: 2494 number of requests in log: 2494 number of expectations: 2494  cpu: 0.42%  mem: 1.345GiB / 5.805GiB
number of requests sent: 2495 number of requests in log: 2495 number of expectations: 2495  cpu: 0.42%  mem: 1.345GiB / 5.805GiB
...
number of requests sent: 2848 number of requests in log: 2848 number of expectations: 2848  cpu: 0.34%  mem: 1.632GiB / 5.805GiB
number of requests sent: 2849 number of requests in log: 2849 number of expectations: 2849  cpu: 0.34%  mem: 1.632GiB / 5.805GiB
number of requests sent: 2850 number of requests in log: 2850 number of expectations: 2850  cpu: 0.34%  mem: 1.632GiB / 5.805GiB
number of requests sent: 2851 number of requests in log: 2851 number of expectations: 2851  cpu: 0.34%  mem: 1.632GiB / 5.805GiB
number of requests sent: 2852 number of requests in log: 2852 number of expectations: 2852  cpu: 0.34%  mem: 1.632GiB / 5.805GiB
number of requests sent: 2853 number of requests in log: 2853 number of expectations: 2853  cpu: 0.34%  mem: 1.632GiB / 5.805GiB
number of requests sent: 2854 number of requests in log: 2854 number of expectations: 2854  cpu: 0.34%  mem: 1.632GiB / 5.805GiB
number of requests sent: 2855 number of requests in log: 2855 number of expectations: 2855  cpu: 0.34%  mem: 1.632GiB / 5.805GiB
number of requests sent: 2856 number of requests in log: 2856 number of expectations: 2856  cpu: 0.34%  mem: 1.632GiB / 5.805GiB
number of requests sent: 2857 number of requests in log: 2857 number of expectations: 2857  cpu: 0.34%  mem: 1.632GiB / 5.805GiB
number of requests sent: 2858 number of requests in log: 2858 number of expectations: 2858  cpu: 0.34%  mem: 1.632GiB / 5.805GiB
number of requests sent: 2859 number of requests in log: 2859 number of expectations: 2859  cpu: 0.34%  mem: 1.632GiB / 5.805GiB^C
```

I will try running the test again with a smaller memory given to the docker container, this test used the default value from docker.

**Comment by jamesdbloom** [comment]
I also just noticed that distroless have release a Java 17 version so I'm upgrading to that as it works better with container memory limits.  They changed the tagging format so I never found it before.

**Comment by busches** [comment]
The issue is the number of starting requests in the JSON file, not adding them latter. I have examples of adding in a few slowly caps the number of expectations and adding in 500 prevented just about any, but I have since proved its not a memory issue.

**Comment by jamesdbloom** [comment]
I understand your tests show that but I'm not sure how that could be the case given the initialiser JSON file is a very thin wrapper around the same logic used by the API.  I've found a few things in the code and reduced the memory footprint of logs due to under certain circumstance a callback not being cleared which due to closure behaviour of the JVM used a lot of memory.  I'll push an improved soon and look into the initialise file issue, although to me that doesn't feel likely.

**Comment by busches** [comment]
I recreated it and detailed it here: https://github.com/mock-server/mockserver/issues/1171#issuecomment-1035050404 🤷

**Comment by jamesdbloom** [comment]
I'm not sure what you're using a custom docker container, because you could just map a volume in and use the normal MockServer container?

With the changes I've committed above I have launched MockServer with an JSON initialiser with 500 expectations in it and added another 500 without any issues at all.  

I have pushed those changes as the latest maven SNAPSHOT and I'll trigger the docker snapshot build again now.

**Comment by busches** [comment]
> I'm not sure what you're using a custom docker container, because you could just map a volume in and use the normal MockServer container?

Mainly to build the JSON file, we keep it split into multiples as one file with 300+ expectations is not maintainable.

> With the changes I've committed above I have launched MockServer with an JSON initialiser with 500 expectations in it and added another 500 without any issues at all.  
> 
> I have pushed those changes as the latest maven SNAPSHOT and I'll trigger the docker snapshot build again now.

We'll give it a shot next week, thanks.

**Comment by busches** [comment]
Still seeing the issue of expectations being forgotten. Memory file now has a lot more entries, but you can still see the number of `expectationsSize` decrease while heap used vs allocation/committed is drastically different (2GB free of heap).

Here's a csv with two runs, we start with 279 expectations from the initial json and as we make calls, we lose them.
[memoryUsage_2022-03-26.csv](https://github.com/mock-server/mockserver/files/8355881/memoryUsage_2022-03-26.csv)

This was the docker image used: `docker.io/mockserver/mockserver:mockserver-snapshot@sha256:c663f559fc143b585fac752adb46d5b2a3b1abcb8639933fe9c5ca8897f4ed70`

Please reopen and let me know if there's any other information I can provide to help debug.

**Comment by busches** [comment]
I can also no longer change the log level using an environment variable or configuration variable:

```
2022-03-26 16:45:54 5.13.1-SNAPSHOT INFO Reading properties from property file [/config/mockserver.properties]:
2022-03-26T16:45:54.689342934Z   mockserver.logLevel = ERROR
2022-03-26T16:45:54.689346816Z   mockserver.memoryUsageCsvDirectory = "/config2"
2022-03-26T16:45:54.689349508Z 
2022-03-26 16:45:55 5.13.1-SNAPSHOT INFO logger level is INFO, change using:
2022-03-26T16:45:55.607687055Z  - 'ConfigurationProperties.logLevel(String level)' in Java code,
2022-03-26T16:45:55.607695916Z  - '-logLevel' command line argument,
2022-03-26T16:45:55.607699435Z  - 'mockserver.logLevel' JVM system property or,
2022-03-26T16:45:55.607702052Z  - 'mockserver.logLevel' property value in 'mockserver.properties' 
2022-03-26T16:45:55.608229975Z 2022-03-26 16:45:55 5.13.1-SNAPSHOT INFO creating expectation:
```

**Comment by busches** [comment]
Swapping to logLevel=ERROR reduces whatever is logged enough that we lose no expectations, but that's not ideal. The main thing I notice is the size of the event log is drastically smaller, not sure what goes into that, but maybe that's part of the reason we're seeing issues? Event log is full?
[memoryUsage_2022-03-26_error.csv](https://github.com/mock-server/mockserver/files/8355909/memoryUsage_2022-03-26_error.csv)

Even though we see the expectations size went from 271 to 236 here, all calls we made were successful and we were able to verify them. 🤷

**Comment by busches** [comment]
In our debugging of it we set `MOCKSERVER_MAX_LOG_ENTRIES: 5000`, moving to that 50k, we can pass at `INFO` level, we still see it hit over 50k, which seems really odd, but it works if I also pass in `"-Xmx4048m", "-Xms2024m"` to increase the heap size. 
[memoryUsage_2022-03-26_info_heap_increased_5k.csv](https://github.com/mock-server/mockserver/files/8355935/memoryUsage_2022-03-26_info_heap_increased_5k.csv)

Removing our specific heap options and the max size, a run gives us a max of 18617 out of the box and we are not able to pass our tests successfully.
[memoryUsage_2022-03-26_info_no_heap_no_max.csv](https://github.com/mock-server/mockserver/files/8355937/memoryUsage_2022-03-26_info_no_heap_no_max.csv)

Adding back a max log entries of 50k, we again pass, but I'm confused on why it defaults to 18k when it can clearly handle 50k. [memoryUsage_2022-03-26_info_no_heap_50kevents.csv](https://github.com/mock-server/mockserver/files/8355938/memoryUsage_2022-03-26_info_no_heap_50kevents.csv)

At this point I'm curious how many events it really wants to log at info, as we only saw 497 at `ERROR`, so I set it to 500k and it got up to 64k.
[memoryUsage_2022-03-26_info_no_heap_500kevents.csv](https://github.com/mock-server/mockserver/files/8355940/memoryUsage_2022-03-26_info_no_heap_500kevents.csv)

Keeping the final settings of INFO, 500k logs, and not setting heap, but running against 5.13.0, and we do not pass. 
[memoryUsage_2022-03-26_info_no_heap_500kevents_failed.csv](https://github.com/mock-server/mockserver/files/8355948/memoryUsage_2022-03-26_info_no_heap_500kevents_failed.csv)
(Note: 5.13.0 does not start if you ask it to log the csv, unless the csv already exists, but that appears to be resolved in the latest snapshot, so I won't make a new issue).

So to recap, I think we can say the original issue is fixed, ONLY if we override the  `MOCKSERVER_MAX_LOG_ENTRIES` when running as `INFO`. Why does MockServer default to 18k max entries when it can support over 64k?
